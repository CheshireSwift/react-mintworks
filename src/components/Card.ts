import { Fragment, ReactNode } from 'react';
import { div, h1, h } from 'react-hyperscript-helpers';
import * as R from 'ramda';
import emoji from 'node-emoji';

import renderEffect from './renderEffect';

export enum CardType {
  Culture,
  Production,
  Utility,
  Deed
}
type CardConfig = {
  name: string;
  cost: number;
  picture: string;
  effect: React.ReactNode;
  score: number;
  type: CardType;
};
type CardProps = {
  card: null | CardConfig;
};

const typeConfig: { [T in CardType]: { icon: string; color: string } } = {
  [CardType.Culture]: { icon: 'herb', color: 'yellowgreen' },
  [CardType.Production]: { icon: 'gear', color: 'indianred' },
  [CardType.Utility]: { icon: 'wrench', color: 'gold' },
  [CardType.Deed]: { icon: 'scroll', color: 'paleturquoise' }
};

const TypeDisplay = (props: { type: CardType }) =>
  div(
    {
      style: { position: 'absolute', top: '1rem', right: '1rem', fontSize: 28 }
    },
    emoji.get(typeConfig[props.type].icon)
  );

const CostDisplay = (props: { cost: number }) =>
  div(
    {
      style: {
        backgroundColor: '#ffffffaa',
        fontSize: 32,
        letterSpacing: -16,
        paddingRight: 16,
        paddingTop: '0.5rem',
        paddingBottom: '0.5rem'
      }
    },
    R.repeat(emoji.get('white_circle'), props.cost)
  );

const EffectDisplay = (props: { effect: ReactNode }) =>
  div({ style: { height: '2rem' } }, renderEffect(props.effect));

const ScoreDisplay = (props: { score: number }) =>
  div(R.repeat(emoji.get('star'), props.score));

const CardFace = ({ name, type, cost, picture, effect, score }: CardConfig) =>
  h(Fragment, [
    h1({ style: { margin: '1rem 0 0' } }, name),
    h(TypeDisplay, { type }),
    h(CostDisplay, { cost }),
    div({ style: { fontSize: 72 } }, picture),
    h(EffectDisplay, { effect }),
    h(ScoreDisplay, { score })
  ]);

const CardBack = () =>
  div(
    {
      style: {
        margin: 'auto',
        fontSize: 192,
        opacity: 0.7,
        textShadow: '16px 8px gray'
      }
    },
    emoji.get('white_circle')
  );

export const Card = ({ card }: CardProps) =>
  div(
    {
      style: {
        textAlign: 'center',
        borderRadius: 4,
        border: '1px solid black',
        backgroundColor: card ? typeConfig[card.type].color : 'paleturquoise',
        height: '20rem',
        width: '15rem',
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'space-between',
        paddingBottom: '1rem',
        position: 'relative',
        fontSize: 'larger'
      }
    },
    [card ? h(CardFace, card) : h(CardBack)]
  );

export default Card;
