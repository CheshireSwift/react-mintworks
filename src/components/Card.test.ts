import { render } from 'enzyme';
import { h } from 'react-hyperscript-helpers';
import emoji from 'node-emoji';

import { Card, CardType } from './Card';

const mintIcon = emoji.get('white_circle');
const starIcon = emoji.get('star');
const utilityIcon = emoji.get('wrench');

describe('Card component', () => {
  const mineConfig = {
    name: 'Mine',
    cost: 2,
    picture: '⛏️',
    effect: '*Upkeep:* Gain ()',
    score: 2,
    type: CardType.Production
  };

  const itDisplaysThe = (field: string, value: any) => {
    it(`displays the ${field}`, () => {
      expect(render(h(Card, { card: mineConfig })).text()).toContain(value);
    });
  };

  it('matches the face snapshot', () => {
    expect(render(h(Card, { card: mineConfig }))).toMatchSnapshot();
  });

  it('matches the back snapshot', () => {
    expect(render(h(Card, { card: null }))).toMatchSnapshot();
  });

  itDisplaysThe('name', mineConfig.name);
  itDisplaysThe('picture', mineConfig.picture);

  it(`displays the cost as ${mintIcon}`, () => {
    expect(
      render(h(Card, { card: { ...mineConfig, cost: 3 } })).text()
    ).toContain(mintIcon + mintIcon + mintIcon);
  });

  it(`displays the score as ${starIcon}`, () => {
    expect(
      render(h(Card, { card: { ...mineConfig, score: 4 } })).text()
    ).toContain(starIcon + starIcon + starIcon + starIcon);
  });

  describe('effect text', () => {
    it('adds strong tags for asterisk delimited bold sections in the effect', () => {
      const strongTags = render(
        h(Card, {
          card: { ...mineConfig, effect: '*bold1* not bold *bold2*' }
        })
      ).find('strong');
      expect(strongTags.length).toBe(2);
      expect(strongTags.text()).toBe('bold1bold2');
    });

    it('displays mints', () => {
      expect(
        render(
          h(Card, {
            card: { ...mineConfig, effect: '*Upkeep:* Gain ()()()()' }
          })
        ).text()
      ).toContain(`Upkeep: Gain ${mintIcon + mintIcon + mintIcon + mintIcon}`);
    });
  });

  describe('type indicator', () => {
    it('displays an emoji for the type', () => {
      expect(
        render(
          h(Card, { card: { ...mineConfig, type: CardType.Utility } })
        ).text()
      ).toContain(utilityIcon);
    });
  });
});
