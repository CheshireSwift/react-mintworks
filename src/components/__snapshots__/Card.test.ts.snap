// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component matches the back snapshot 1`] = `
<div
  style="text-align:center;border-radius:4px;border:1px solid black;background-color:paleturquoise;height:20rem;width:15rem;display:flex;flex-direction:column;justify-content:space-between;padding-bottom:1rem;position:relative;font-size:larger"
>
  <div
    style="margin:auto;font-size:192px;opacity:0.7;text-shadow:16px 8px gray"
  >
    ⚪
  </div>
</div>
`;

exports[`Card component matches the face snapshot 1`] = `
<div
  style="text-align:center;border-radius:4px;border:1px solid black;background-color:indianred;height:20rem;width:15rem;display:flex;flex-direction:column;justify-content:space-between;padding-bottom:1rem;position:relative;font-size:larger"
>
  <h1
    style="margin:1rem 0 0"
  >
    Mine
  </h1>
  <div
    style="position:absolute;top:1rem;right:1rem;font-size:28px"
  >
    ⚙️
  </div>
  <div
    style="background-color:#ffffffaa;font-size:32px;letter-spacing:-16px;padding-right:16px;padding-top:0.5rem;padding-bottom:0.5rem"
  >
    ⚪⚪
  </div>
  <div
    style="font-size:72px"
  >
    ⛏️
  </div>
  <div
    style="height:2rem"
  >
    <strong>
      Upkeep:
    </strong>
     Gain ⚪
  </div>
  <div>
    ⭐⭐
  </div>
</div>
`;
